### ISSUE TITLE
V4.0 and Higher Casts Params to 1st Specified Type using Typing Syntax

### ISSUE THREAD

**Issue by rmccarth-bosch** [robotframework/robotframework#3926]
Hi,
This is my first time ever submitting a bugfix request so please take it easy on me :) FYI, longtime fan of this work... My company has converted to use this exclusively for our automated testing. We leverage custom backend libraries in Python and use typing in params to show the accepted formats.  I ran into this issue when bumping from 3.1.2 -> 4.0 on up.

Consider the below files
Supporting library Lib.py:
```
class Lib:    
    '''
    Does cool stuff
    '''

    def __init__ ( self ):
        pass

    def check_param_as_str ( self, param : (str or list) ):
        '''
        Accepts param as type str or list -- asserts when passed in str
        '''
        if type(param) == str:
            raise AssertionError ("Ran into type<str> when expected type<list>")
        return param
    
    def check_param_as_str_without_typing ( self, param ):
        '''
        Accepts param as type str or list -- asserts when passed in str
        '''
        if type(param) == str:
            raise AssertionError ("Ran into type<str> when expected type<list>")
        return param

```
and test.robot:
```
*** Settings ***
Library  ${CURDIR}/Lib.py

*** Test Cases ***
WITH TYPING: Pass in Str
    ${a_str}  Set Variable  some param
    ${param} =  Run Keyword And Expect Error  Ran into type<str> when expected type<list>  Check Param as Str  ${a_str}
    Log  ${param}

WITH TYPING: Pass in List (SHOULD Pass)
    ${a_list} =  Create List  more params
    ${param} =  Check Param as Str  ${a_list}
    Log  ${param}

WITHOUT TYPING: Pass in Str
    ${a_str}  Set Variable  some param
    ${param} =  Run Keyword And Expect Error  Ran into type<str> when expected type<list>  Check Param as Str Without Typing  ${a_str}
    Log  ${param}

WITHOUT TYPING: Pass in List
    ${a_list} =  Create List  more params
    ${param} =  Check Param as Str Without Typing  ${a_list}
    Log  ${param}

```
On 3.1.2  < version < 4.0, the test cases fully pass and typing works as expected.

On version >= 4.0, the test case 'WITH TYPING: Pass in List (SHOULD Pass)' fails. It appears that using typing with style `param:(<type> or <type>)` will look to cast the param to the first provided <type>.  Whereas style `param: (<type>, <type>)` will not effect the param.

It should be noted, when running this code exclusively against Python, latest, the behavior acts like 3.1.2  < version < 4.0.

I am going to try to attach some relevant screenshots / files. Thank you!

**Comment by pekkaklarck** [comment]
First of all, thanks for kind words! I've had pretty rough few days and compliments are appreciated even more than normally. This is a pretty interesting issue because there's a bug in your example but there's also a bug in Robot related to this. I'll explain both below.

The bug in your example is that you cannot specify that argument would have multiple possible types like `param: (str or list)`. Although it reads like it could work, the type hint is evaluated in Boolean context so that its value is `str` if `str` is considered true and `list` otherwise. Since types are true this is exactly same as just using `param: str`. It's easy to demonstrate that on the Python interpreter:

```python
>>> def func(arg: (str or list)):
...     pass
... 
>>> func.__annotations__
{'arg': <class 'str'>}
```

The right way to specify that an argument has multiple possible values is using [typing.Union](https://docs.python.org/3/library/typing.html#typing.Union) like `Union[str, list]`. With Robot Framework you can actually also use a tuple like `(str, list)`, but with Python 3 code `Union` is generally recommended. 

Why the code worked with versions prior to RF 4.0 is that earlier type conversion was done only if the given argument was a string. Thus when you effectively had type hint like `param: str` and called it with a list, there was no conversion.

The bug in Robot is that it does unnecessary conversion if an argument has multiple possible types and the given argument is one of them. Instead of passing the argument as-is, it converts it to the first possible type. Thus if you'd have type hint `Union[str, list]` and call it with a list, the list is converted to a string. This will be fixed in RF 4.0.1 (#3897) that I hopefully get released still today. The fix is already in RF 4.0.1 beta 1 so you can install that if you want to test this already before. In this particular case a workaround that works also with RF 4.0 would be using `Union[list, str]`.

Could you test does `Union` work for you as explained above?

**Comment by rmccarth-bosch** [comment]
Hi @pekkaklarck, No problem at all and I am so sorry to hear about the rough times you are experiencing. I hope things calm down for you. Thank you so much for the detailed and kind response, I really appreciate the extra insight into both how the Python and Robotframework handles this.

Those are some very interesting points and  I did not realize that was an invalid syntax we were using.  In fact, after talking to my system architect we believe that the `Union` will work for us.  

While the tuple style of `(str, list)` may work with the framework, our libraries are also duplicated to work with external Python scripts... The Union example you provided will be perfect for our use case. Thanks again!

**Comment by pekkaklarck** [comment]
Yeah, `Union[int, str]` is typically the way to go, at least until Python 3.10 makes it possible to write that as `int | str`. Notice how that's actually pretty close to what you tried to use. Using a tuple may work better with the `@keyword` decorator, though, and that's the only Python 2 compatible way.

**Comment by pekkaklarck** [comment]
Since everything seems to work as expected I close this issue, but you can comment this or submit a new one if you encounter problems. Just remember to upgrade to the just released RF 4.0.1 if you `Union` to get a fix to #3897.
